sample-authz/src/main/java/net/trajano/ms/example/authz/TokenResource.java

```diff
@@ -0,0 +1,177 @@
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+
+@Api
+@Configuration
+@Component
+@Path("/token")
+@PermitAll
+public class TokenResource {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TokenResource.class);
+
+    @Autowired
+    private TokenGenerator tokenGenerator;
+
+    @Autowired
+    private ClientValidator clientValidator;
+
+    @Value("${realmName:client_credentials}")
+    private String realmName;
+
+    private Clients clients;
+
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    /**
+     * Performs client credential validation then dispatches to the appropriate
+     * handler for a given grant type.
+     */
+    @POST
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public OAuthTokenResponse dispatch(
+        @ApiParam(allowableValues = "refresh_token, authorization_code") @FormParam("grant_type") @NotNull String grantType,
+        @FormParam("code") String code,
+        @FormParam("assertion") String assertion,
+        @FormParam("refresh_token") String refreshToken,
+        @HeaderParam(HttpHeaders.AUTHORIZATION) @NotNull String authorization) {
+
+        final String[] clientCredentials = HttpAuthorizationHeaders.parseBasicAuthorization(authorization);
+
+        if (clientCredentials == null) {
+            throw OAuthTokenResponse.unauthorized("unauthorized_client", "Missing credentials", String.format("Basic realm=\"%s\", encoding=\"UTF-8\"", realmName));
+        }
+        final String clientId = clientCredentials[0];
+        if (!clientValidator.isValid(grantType, clientId, clientCredentials[1])) {
+            throw OAuthTokenResponse.unauthorized("unauthorized_client", "Unauthorized client", String.format("Basic realm=\"%s\", encoding=\"UTF-8\"", realmName));
+        }
+
+        if (GrantTypes.REFRESH_TOKEN.equals(grantType)) {
+            return handleRefreshGrant(refreshToken, clientId);
+        } else if (GrantTypes.AUTHORIZATION_CODE.equals(grantType)) {
+            return handleAuthorizationCodeGrant(code, clientId);
+        } else if (GrantTypes.JWT_ASSERTION.equals(grantType)) {
+            return handleJwtAssertionGrant(assertion, clientId);
+
+        } else {
+            throw OAuthTokenResponse.badRequest("invalid_grant", "Invalid grant");
+        }
+
+    }
+
+    /**
+     * Maximum life of a JWT token. Past that period, it is expected to no
+     * longer be used.
+     */
+    @Value("${token.jwtMaximumLifetime:86400}")
+    private int jwtMaximumLifetimeInSeconds;
+
+    @Autowired
+    private TokenCache tokenCache;
+
+    @Value("${issuer}")
+    private URI issuer;
+
+    @Autowired
+    private InternalClaimsBuilder internalClaimsBuilder;
+
+    private OAuthTokenResponse handleJwtAssertionGrant(final String assertion,
+        final String clientId) {
+
+        if (assertion == null) {
+            throw OAuthTokenResponse.badRequest("invalid_request", "Missing Assertion");
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "invalid_request" 4 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)